### PR TITLE
Merge dev branch into main

### PR DESCRIPTION
**Pull request by DnPlas**
All changes in this branch have already been reviewed and approved in #9 and #10. This is an attempt to merge those commits into main.

### CODE REVIEW

File: charms/kserve-controller/src/charm.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    "src/templates/crd_manifests.yaml.j2",
+    "src/templates/auth_manifests.yaml.j2",
+    "src/templates/serving_runtimes_manifests.yaml.j2",
+    "src/templates/webhook_manifests.yaml.j2",
+    "src/templates/configmap_manifests.yaml.j2",
+]
+CONFIG_FILE = "src/config/inference-config.j2"
+
+
+class CheckFailed(Exception):
```
Comment (i-chvets) [inline charms/kserve-controller/src/charm.py:47]: There is chisme ErrorWithStatus that can be used instead.

File: charms/kserve-controller/src/charm.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+                }
+            }
+        )
+
+    def _on_kserve_controller_ready(self, event):
+        """Define and start a workload using the Pebble API.
+
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
```
Comment (i-chvets) [inline charms/kserve-controller/src/charm.py:146]: I think we should check container connection here before uploading anything:
`self.container.can_connect()` will return `True` if container available.

File: charms/kserve-controller/src/charm.py (hunk 1 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:47]: True, I'll remove it.

File: charms/kserve-controller/src/charm.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        # TODO determine status checking if rbac proxy is also up
+        self.unit.status = ActiveStatus()
+
+    def _on_kube_rbac_proxy_ready(self, event):
+        """Define and start a workload using the Pebble API.
+
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            update_layer(
```
Comment (i-chvets) [inline charms/kserve-controller/src/charm.py:174]: Same here with container connection check.

File: charms/kserve-controller/src/charm.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            log.warning(f"Failed to delete resources, with error: {e}")
+            raise e
+        self.unit.status = MaintenanceStatus("K8s resources removed")
+
+    def gen_certs(self, namespace, service_name):
+        if Path("/run/cert.pem").exists():
+            log.info("Found existing cert.pem, not generating new cert.")
+            return
+    
+        Path("/run/ssl.conf").write_text(
```
Comment (i-chvets) [inline charms/kserve-controller/src/charm.py:218]: Just a though. In my re-write I checked in `ssl.conf.j2` as under `src/templates/` and used it to render the real `ssl.conf`
Kinda makes it more uniform and removes long strings from code.

File: charms/kserve-controller/tests/integration/test_charm.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+        assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
+
+
+@pytest.mark.abort_on_fail
+async def test_application_is_up(ops_test: OpsTest):
+    status = await ops_test.model.get_status()  # noqa: F821
```
Comment (i-chvets) [inline charms/kserve-controller/tests/integration/test_charm.py:46]: Doc string. Linting will fail.

File: charms/kserve-controller/tests/unit/test_charm.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
+# Copyright 2022 Canonical Ltd.
```
Comment (i-chvets) [inline charms/kserve-controller/tests/unit/test_charm.py:1]: Doc strings in all tests. Linting will fail without those.

File: charms/kserve-controller/tests/unit/test_charm.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    mocked_resource_handler.lightkube_client = MagicMock()
+    yield mocked_resource_handler.lightkube_client
+
+
+@pytest.fixture()
+def mocked_gen_certs(mocker):
+    """Yields a mocked gen_certs."""
+    yield mocker.patch("charm.KServeControllerCharm.gen_certs")
+
+
```
Comment (i-chvets) [inline charms/kserve-controller/tests/unit/test_charm.py:80]: Doc strings in all tests from this line to the end. Linting will fail without those.

File: charms/kserve-controller/src/charm.py (hunk 3 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:174]: chisme will [check](https://github.com/canonical/charmed-kubeflow-chisme/blob/main/src/charmed_kubeflow_chisme/pebble/_update_layer.py#L22) that

File: charms/kserve-controller/src/charm.py (hunk 1 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:47]: [Removed](https://github.com/canonical/kserve-operators/blob/KF-803-kserve-refactor/charms/kserve-controller/src/charm.py)

File: charms/kserve-controller/src/charm.py (hunk 2 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:146]: I have added the [check](https://github.com/canonical/kserve-operators/blob/KF-803-kserve-refactor/charms/kserve-controller/src/charm.py#L132)

File: charms/kserve-controller/src/charm.py (hunk 4 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:218]: This suggestion will be included in the refactor. You can see it [here](https://github.com/canonical/kserve-operators/blob/KF-803-kserve-refactor/charms/kserve-controller/src/charm.py#L216)

File: charms/kserve-controller/tests/integration/test_charm.py (hunk 5 above)
Comment (DnPlas) [reply charms/kserve-controller/tests/integration/test_charm.py:46]: Will add all docstrings in refactor.

**Pull request merged**